Sources/FrontEnd/TypeChecking/Rewriting/RewritingSystem.swift

```diff
@@ -0,0 +1,281 @@
+        let x = Term(u[..<p])
+        let v = rules[r].rhs
+        let z = u[(p + rules[r].lhs.count)...]
+        return reduce(x + v + z)
+      }
+    }
+    return u
+  }
+
+  /// Inserts the given rule using `compareOrder` to order pairs of terms.
+  ///
+  /// A rule *u => v* is notionally contained in a rewriting system if that system contains a set
+  /// of rules capable of rewriting *u* to *v*. This set may not contain *u => v* itself.
+  ///
+  /// The return value is `(true, new)` if `r` is not already notionally contained in the system,
+  /// where `new` is the identifier of a newly inserted rule that is encoding `r`. Otherwise, the
+  /// return value is `(false, old)` where `old` is the identifier of a rule encoding `r`.
+  ///
+  /// - Precondition: The source of the rule is ordered after its target.
+  @discardableResult
+  mutating func insert(
+    _ r: RewritingRule<Term>,
+    orderingTermsWith compareOrder: (Term, Term) -> StrictOrdering
+  ) -> (inserted: Bool, ruleAfterInsertion: RuleID) {
+    precondition(compareOrder(r.lhs, r.rhs) == .descending, "invalid rewriting rule")
+    precondition(!r.isSimplified)
+
+    // If the source of the rule isn't associated with any other rule yet, inserts the rule and
+    // return `(true, i)` where `i` is the position of the rule in `rules`. Otherwise, return
+    // `(false, j)` where `j` is the position of a rule sharing the same source.
+    let result = modify(&termToRule[r.lhs]) { (q) in
+      if let old = q {
+        return (inserted: false, ruleAfterInsertion: old)
+      } else {
+        q = rules.count
+        rules.append(r)
+        return (inserted: true, ruleAfterInsertion: q!)
+      }
+    }
+
+    // Nothing more to do if the rule was inserted.
+    if result.inserted {
+      return result
+    }
+
+    // Otherwise, update the system to notionally contain the rule.
+    let old = result.ruleAfterInsertion
+    switch compareOrder(r.rhs, rules[old].rhs) {
+    case .equal:
+      return result
+
+    case .descending:
+      return insert(.init(r.rhs, rules[old].rhs), orderingTermsWith: compareOrder)
+
+    case .ascending:
+      // Let the u1 => v1 and u2 => v2 be the old and new rules, respectively. Remove u1 => v1 from
+      // the system and add v1 => v2.
+      rules[old].raiseFlags(.isRightSimplified)
+      insert(.init(rules[old].rhs, r.rhs), orderingTermsWith: compareOrder)
+
+      // Add u2 => v2.
+      let q = rules.count
+      termToRule[r.lhs] = q
+      rules.append(r)
+      return (inserted: true, ruleAfterInsertion: q)
+    }
+  }
+
+  /// Turns `self` into a terminating and confluent rewriting system, using `compareOrder` to order
+  /// pairs of terms.
+  ///
+  /// This method uses Knuth-Bendix completion algorithm to transform `self` into a terminating
+  /// confluent system. The completion procedure is semi-decidable: it returns `true` if it
+  /// succeeds or `false` if it suspects that it won't terminate.
```

Comment:
```suggestion
  /// Turns `self` into a terminating and confluent rewriting system using `compareOrder` to order
  /// pairs of terms and returns `true`, or returns `false` indicating that the transformation 
  /// steps reached a fixed limit without succeeding.
  ///
  /// This method uses the semi-decidable Knuth-Bendix completion algorithm, thus the
  /// transformation step limit.  If the transformation fails and returns `false`, `self` is
  /// arbitrarily mutated.
```

The idea of this suggestion is to capture the entire contract in the summary, which is almost always possible.

This is the sort of thing for which I might use a type transformation, and return an optional of the transformed type:

```swift
func completed(orderingTermsWith compareOrder: (Term, Term) -> StrictOrdering) -> CompletedSystem?
```

I don't know if that fits into the rest of your world though.